### PR TITLE
Fix exceptions not being caught

### LINKED ISSUE #1297: Tests fail on ppc64le
As reported in #922.

### PR DESCRIPTION
**Pull request by brunoalr**
Adding -fexceptions to both torch and pytorch C/C++ builds fixes tests
not passing.

Related to https://github.com/pytorch/pytorch/pull/1309

Closes #1297

### CODE REVIEW

**Comment by brunoalr** [comment]
I opened a new PR because I pushed with --force to my PR branch.

**Comment by brunoalr** [comment]
Rebasing and solving conflicts

**Comment by brunoalr** [comment]
@soumith can you ping your buildbot for this PR as well?

**Comment by soumith** [comment]
i need to fix this PR, your merge added back stuff that was removed. No worries i'll do it.

File: setup.py [hunk 1]
```diff
@@ -396,6 +396,10 @@ def make_relative_rpath(path):
 # Declare extensions and package
 ################################################################################
 
+# nvcc does not support '-fexceptions'
```
Comment (soumith) [inline setup.py]: this change shouldn't be necessary right? 
the THCUNN extension does not call NVCC. It just autogenerates C++ wrappers and would only call CXX.

**Review by soumith — CHANGES_REQUESTED** [review]
I have a question (see inline).
From my expectation, the setup.py change is not needed, but can you test it on PPC64 and let me know?

The changes to build_all.sh look good.

File: torch/lib/build_all.sh [hunk 2]
```diff
@@ -28,35 +28,39 @@ cd "$(dirname "$0")/../.."
...
               -DTorch_FOUND="1" \
               -DCMAKE_INSTALL_PREFIX="$INSTALL_DIR" \
-              -DCMAKE_C_FLAGS="$C_FLAGS" \
-              -DCMAKE_CXX_FLAGS="$C_FLAGS $CPP_FLAGS" \
-              -DCMAKE_EXE_LINKER_FLAGS="$LDFLAGS" \
-              -DCMAKE_SHARED_LINKER_FLAGS="$LDFLAGS" \
-              -DCUDA_NVCC_FLAGS="$C_FLAGS" \
+              -DCMAKE_C_FLAGS="$BUILD_C_FLAGS" \
+              -DCMAKE_CXX_FLAGS="$BUILD_C_FLAGS $CPP_FLAGS" \
+              -DCUDA_NVCC_FLAGS="$BASIC_C_FLAGS" \
```
Comment (apaszke) [inline torch/lib/build_all.sh]: Why did you move LDFLAGS into CFLAGS?

Comment (brunoalr) [inline torch/lib/build_all.sh]: Sorry, I made a mistake when rebasing this branch. @soumith is that what you were talking about?

Comment (soumith) [inline torch/lib/build_all.sh]: yep. i was gonna fix that, pending your answer on `setup.py`. You can fix both if you're making changes :)

File: setup.py (hunk 1 above)
Comment (brunoalr) [inline setup.py]: Actually, it is necessary for that very reason. The idea is to pass `-fexceptions` to all the modules that **does not** call NVCC (since it doesn't support the flag).

**Comment by brunoalr** [comment]
@soumith @apaszke fixed! :)

**Review by soumith — APPROVED** [review]

**Comment by soumith** [comment]
Thanks Bruno!
